lib/spack/spack/test/concretize.py

```diff
@@ -2295,7 +2295,7 @@ def test_explicit_splices(
 
         spec = spack.spec.Spec("hdf5 ^zmpi").concretized()
 
-        assert spec.satisfies(f"^mpich/{mpich_spec.dag_hash()}")
+        assert spec.satisfies(f"^mpich@{mpich_spec.version}")
```

Comment:
I would also test the `build_spec` here to make sure it's exactly what's expected.

Is the issue that `fake` (which `zmpi` depends on) or some runtime lib is getting added in with an intransitive splice on the compiler deps branch?  Might also be good to isolate that somehow for these tests (or test both with and without the runtime libs).

Reply:
Agree on testing the build_spec

(The compiler runtime libs will be spliced in, not `fake`)

I don't think we need to add complexity to this splice for that, this is just testing that the splice occurred. We have two complex tests in `spec_semantics` that test a pair of pretty robustly convoluted specs transitive and intransitive.

Reply:
Aren't we testing the build spec already:
```python
assert spec.build_spec.dependencies(name="zmpi", deptype="link")
assert not spec.build_spec.satisfies(f"^mpich/{mpich_spec.dag_hash()}")
```
? This ensures that the build spec is using `zmpi` and not `mpich`.

Reply:
Done in https://github.com/spack/spack/pull/46914/commits/14c188ea07acea5525097583897a0cba5cc7f5ec